ADR/0016-integration-service-promotion-logic.md

```diff
@@ -0,0 +1,86 @@
+# 16. Promotion logic in the Integration Service
+
+Date Documented: 2023-02-08
+
+## Status
+
+In Review
+
+## Context
+
+Before the merge of HACBS & AppStudio, the AppStudio build-service created the
+ApplicationSnapshot and promoted the content to the user's lowest environment as soon as
+the build was completed and once the environment was built.
+The integration-service fulfilled the same role for HACBS in addition to running the 
+IntegrationTestScenario tests first before promoting.
+
+Here, we are documenting the decision to consolidate and move the logic at one place to avoid duplicacy
+on the actions.
+
+Note: This functionality has now been completely dropped from the build-service side after the merger.
+
+### What is promotion logic ?
+
+The movement of the container image values of components, between environments, along a prescribed 
+promotion graph (e.g. ‘dev’ env -> ‘staging’ env -> ‘prod’ env).
+
+Once the Snapshot is tested and verified successfully, its contents will then be deployed to the user's 
+lowest environments as per the information concluded from the user’s namespace. 
+
+The “lowest environments” are the Environments with no explicit parentEnvironment set.
+
+## Decision
+
+Consolidate the promotion logic for both HACBS & AppStudio and move it to the integration-service after 
+the merger of Stonesoup.
+
+![](../diagrams/ADR-0016/promotion-logic.jpg)
+
+Three steps to follow in the integration-service to validate and promote the content of the successful Snapshot. 
+It:
+* Ensures that the ContainerImage in the Global Candidate List(GCL) is being updated when the Snapshot 
+  passes all the integration tests.
+* Ensure all pipeline Releases associated to the Snapshot and the Application's ReleasePlans exist. 
+  It will create new Releases for each ReleasePlan if it doesn’t exist.
+* Ensure all SnapshotEnvironmentBindings(SEB) for non-ephemeral lowest environments point to the newly 
+  constructed Snapshot. It will create new ones for each of the environments if the bindings don't already exist.
+
+Note: In case of no IntegrationTestScenario defined, the testing of the Snapshot will be skipped and the 
+outcome will be considered as passed followed by the same promotion process defined in the diagram.
```

Comment:
By current design, the `IntegrationTestScenario` can specify which environment it wants to use. The Integration service then creates (by way of gitops service) an ephemeral copy of the specified environment and provides the credentials to the Integration pipeline run so it can execute tests on the newly provisioned environment.

Reply:
But that wouldn't be the same as running a smoke test after a promotion to that environment. It would still be using a new ephemeral environment instead of deploying/upgrading the currently running environment on the Snapshot's way to the highest environment. 

I was trying to think about that latter challenge of being able to introduce gating into the promotion process.

Reply:
Hm, I'm not seeing that happening as part of the Integration testing logic, because of the parallel testing of different snapshots and running promotions after that.
This idea reads more to me like mid-promotion logic, right? That would require an additional loop to be added to the Integration service.

The `IntegrationTestScenario` does have the (currently underutilized) [context field](https://github.com/redhat-appstudio/integration-service/blob/main/api/v1alpha1/integrationtestscenario_types.go#L41) where the user could specify that they want to run tests in specific cases. Implementing something based on that wouldn't be out of the question, but it would require some non-trivial work to be done to support it.

Reply:
... and a new issue, etc. ;) 

Yes, it is a mid-promotion logic to run tests against a running environment and not provision a new one freshly.